### PR TITLE
Add Record Type mapping for Bulk Data tasks

### LINKED ISSUE #1270: Make generate mappings task record-type aware
EDA and other managed packages rely heavily on record types. Generate Mappings is not currently record-type aware, which eliminates a strong and necessary datapoint for data flows.

### PR DESCRIPTION
**Pull request by davidmreed**
# Critical Changes

# Changes

- Add support for mapping Record Types between orgs (by Developer Name) during bulk data extract and load.
- Add support for Record Type mapping in bulk mapping generator.
- Corrects a problem resulting in `generate_dataset_mapping` outputting an incorrectly sequenced mapping.
- Adds new documentation for bulk data tasks

# Issues Closed

- Fixes #1270

### CODE REVIEW

**Comment by jstvz** [comment]
Unrelated to your changes: Do we still need `unicodecsv`?

**Comment by davisagli** [comment]
@jstvz good question -- probably not

**Comment by davidmreed** [comment]
I don't think so. I've used Python 3 `csv` with Unicode data extensively.

File: cumulusci/tasks/bulkdata/utils.py [hunk 1]
```diff
@@ -128,6 +130,29 @@ def _sql_bulk_insert_from_csv(self, conn, table, columns, data_file):
                 conn.execute(table.insert().values(rows))
         self.session.flush()
 
+    def _create_record_type_table(self, table_name):
+        rt_map_model_name = "{}Model".format(table_name)
```
Comment (Br4nd0R) [inline cumulusci/tasks/bulkdata/utils.py]:
```suggestion
        rt_map_model_name = f"{table_name}Model"
```

File: cumulusci/tasks/bulkdata/utils.py [hunk 2]
```diff
@@ -128,6 +130,29 @@ def _sql_bulk_insert_from_csv(self, conn, table, columns, data_file):
...
+        self.models[table_name] = type(rt_map_model_name, (object,), {})
+        rt_map_fields = [
+            Column("record_type_id", Unicode(18), primary_key=True),
+            Column("developer_name", Unicode(255)),
+        ]
+        rt_map_table = Table(table_name, self.metadata, *rt_map_fields)
+        mapper(self.models[table_name], rt_map_table)
+
+    def _extract_record_types(self, sobject, table, conn):
+        self.logger.info("Extracting Record Types for {}".format(sobject))
```
Comment (Br4nd0R) [inline cumulusci/tasks/bulkdata/utils.py]:
```suggestion
        self.logger.info(f"Extracting Record Types for {sobject}")
```

**Comment by davidmreed** [comment]
@Br4nd0R I converted the bulk data module to fstrings throughout, added documentation, and addresseed @prescod's suggestion that the mapping generator should refrain from lower-casing Salesforce API names.

**Pull request merged**